Document secrets

mstanleyjones opened the pull request:
### Describe the proposed changes

Add docs for using secrets, addresses #529.

### Unreleased project version

Engine 1.13

### Related issue

#529 

### Related issue or PR in another project

see #529

### Please take a look

@ehazlett @thaJeztah @mrjana @stevvooe and please add anyone else who should review.


<!-- To improve this template, edit .github/PULL_REQUEST_TEMPLATE.md. -->

johndmulhausen commented:
Deploy preview failed.

Built with commit 1243ffc60f2903e76ba5164c0048693457212a0f

https://app.netlify.com/sites/docker-docs/deploys/582a634da7e2d21444fb3cf6

stevvooe commented on _data/toc.yaml:
```diff
@@ -219,6 +219,8 @@ toc:
       title: Manage nodes in a swarm
     - path: /engine/swarm/services/
       title: Deploy services to a swarm
+    - path: /engine/swarm/store_sensitive_strings/
```
Why not `secrets`?

mstanleyjones replied:
oh, I suppose.... I just don't think secrets means anything to anyone. :)

diogomonica replied:
Agreed. Let's use sensitive data.

cyli replied:
Non-blocking: it looks like in the rest of the documentation there is a parenthetical "(secrets)" after "manage sensitive strings" - should that also be included here so folks can more easily find it in the TOC if they're specifically looking for the word "secrets"?

stevvooe commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+itself has already been propagated to the swarm node using mutual TLS.
+
+>**Warning**: RAFT data is encrypted in Docker 1.13 and newer. If some of your
+Swarm nodes run an earlier version, the secrets are stored unecrypted in
```
> the secrets __may be__ stored unecrypted in those nodes' RAFT logs.

I'm not sure that we opaquely store secret data in older versions.

@diogomonica @cyli ?

diogomonica replied:
If we're in a situation where there is a 1.12 manager in a 1.13 cluster, and someone adds a secret, the 1.12 manager will write the secrets unencrypted to disk.

The upgrade path should be: upgrade all the managers to 1.13, and then start using secrets.

diogomonica replied:
We should probably add something about how to securely upgrade from a 1.12, or a note that to use secrets securely, make sure that all of the manager nodes are running 1.13

stevvooe replied:
@diogomonica While RAFT messages may be written to disk opaquely, I'm not sure that a manager with 1.12 snapshot structure definition will actually do this. None of our types have opaque field preservation, as far as I can tell. Looks like this was removed in proto3, as described in https://github.com/google/protobuf/issues/272.

> The upgrade path should be: upgrade all the managers to 1.13, and then start using secrets.

I agree and this language is sufficient to protect users against leaks, even if not precisely accurate. We may want to make this stronger and warn users to make sure the _entire_ cluster is on 1.13 before performing cluster mutations.

The only data loss scenario @aaronlehmann and I could work out was if a 1.12 engine was elected leader in a mixed-mode cluster. We may be able to downweight older engines in the master election process to minimize this possibility.

The other scenarios for data loss considered are 1.12 proxying to 1.13 leader, but this is a non-issue, since the JSON API prevents opaque field transfer.

aaronlehmann replied:
> The only data loss scenario @aaronlehmann and I could work out was if a 1.12 engine was elected leader in a mixed-version cluster.

Or if that engine is upgraded to 1.13 and later becomes the leader. It will be missing the data.

stevvooe commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+revoke its access to a given secret at any time.
+
+The Engine managing a given worker node creates a bind mount within
+`/run/secrets/` on the worker node for each secret. From the point of view of
```
@ehazlett `/run/secrets` runs afoul of FHS standards. Do we link to `/var/run` to `/run`?

From https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard:

> In FHS 3.0, /var/run is replaced by /run; a system should either continue to provide a /var/run directory, or provide a symbolic link from /var/run to /run, for backwards compatibility.[10]

I think we just need to make sure that `/run` is not a fork of `/var/run`. Accessing `/var/run/secrets` or `/run/secrets` should be equivalent, recommending the later.

ehazlett replied:
Yes /var/run is linked to /run

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+---
+title: Manage sensitive strings (secrets) for Docker services
+description: How to securely store, retrieve, and use sensitive strings (secrets) in Docker services
```
the use of strings is weird. what about `sensitive data` everywhere?

mstanleyjones replied:
Could you store sensitive data that isn't a string?

diogomonica replied:
String has a very specific meaning for developers and programmers. I think we want to avoid using that term. The term `blob` would technically be more adequate, but I think `sensitive data` reads a lot better.

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+
+## About secrets
+
+In terms of Docker Swarm services, a _secret_ is a string, such as a password,
```
`string` -> `a blob of data`

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+## About secrets
+
+In terms of Docker Swarm services, a _secret_ is a string, such as a password,
+credit card number, SSH private key, SSL certificate, or another piece of
```
Definitely not a credit card number. Lets remove that.

stevvooe commented:
A few nits, but, otherwise, a great doc!

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+In terms of Docker Swarm services, a _secret_ is a string, such as a password,
+credit card number, SSH private key, SSL certificate, or another piece of
+sensitive data that should not be transmitted over a network or stored in a
+Dockerfile or in your application's source code in plain text. Starting in
```
starting with?

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+Docker 1.13, Docker can encrypt and manage these secrets for you. Consider the
+following examples when you might want to manage sensitive strings using Docker:
+
+- Your service relies on a password-protected database, and you want to protect
```
`and you want to securely distribute the credentials to your application containers`

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+  different services running in your swarm.
+- Your application relies upon a SSH private key, and you want to protect the
+  key.
+- You run an application, such as a web server, which needs to use a SSL
```
use TLS instead of SSL

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+  key.
+- You run an application, such as a web server, which needs to use a SSL
+  certificate, and you want to protect the certificate itself.
+- You use a different SSL certificate in development and production environments
```
TLS

diogomonica replied:
Also, not sure exactly what you mean with this usecase.

mstanleyjones replied:
You could have a different certificate but stored with the same secret name, and deploying to devel or production would be the same, even though the actual secret would be different.

diogomonica replied:
Sounds good.

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+  certificate, and you want to protect the certificate itself.
+- You use a different SSL certificate in development and production environments
+  but do not want to change the image or store the SSL certificate in the image.
+- Your application uses infrastructure which is metered and needs a credential,
```
What about just listing the kinds of secrets instead of creating an example for each individual one?
```
Your application needs secrets like:
- API tokens
- SSH private keys
- TLS keys
...
```

mstanleyjones replied:
I really wanted to answer "why would I use this" as well as "What could I use this for"...

diogomonica replied:
I get it. My feedback is that it doesn't seem to add much, and there is a shorter/sweeter way of describing potential use-cases, but I'm not known for writing great documentation :)

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+  populate the secret key with a mock value while you are developing the
+  application, and change it to use the real value in production.
+
+In these and many other situations, managing sensitive strings using Docker
```
sensitive data

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+
+## How Docker manages secrets
+
+When you add a secret to the swarm, Docker sends the secret to each manager
```
Maybe a nit, but we don't send the secret to each manager. We send the secret to a manager, and then the secret gets replicated securely across the remaining managers.

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+## How Docker manages secrets
+
+When you add a secret to the swarm, Docker sends the secret to each manager
+node over a network connection which uses mutual TLS. When a manager receives
```
What about: when a manager receives a secret, it stores it in RAFT, relying on the RAFT logs replication to ensure that your secrets are distributed with the remaining managers, and ensuring your secrets are highly available in the event of a manager failure.

In Docker 1.13 and newer, all the RAFT logs are encrypted using NaCl SecretBox (Salsa20/Poly1305)....

diogomonica replied:
My point is:
- Talk about high availability of secrets due to using raft for replication
- Talk about MTLS for transport of the data
- Talk about how all raft logs are encrypted at rest using NaCl SecretBox (Salsa20/Poly1305).

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+`/run/secrets/` on the worker node for each secret. From the point of view of
+the container, the secret is available as plain text. The `/run/secrets/`
+directory is an in-memory filesystem. When the node stops or leaves the swarm,
+or when the service is stopped or removed, the secrets are no longer accessible.
```
add: and are completely removed from the node's memory.

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+## Read more about `docker secret` commands
+
+Use these links to read about specific commands, or continue to the
+[example about using secrets with a service](store_sensitive_strings.md@example-use-secrets-with-a-service).
```
I like `store sensitive data` instead of `sensitive strings`

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+When evaluating the value of the environment variable,
+first check whether that value corresponds to a file path available within the
+container read the credentail from the file if so. The `mistysj/wordpress` image
+adds the following check to see if the value of `WORDPRESS_DB_PASSWORD` is a
```
We should probably document why we don't support environment variables in the first place. They are less secure than files.

See: https://github.com/docker/docker/pull/9176#issuecomment-99542089

diogomonica commented on engine/swarm/store_sensitive_strings.md:
```diff
@@ -0,0 +1,234 @@
+- [`--secret`](../reference/commandline/service_create.md#create-a-service-with-secrets) flag for `docker service create`
+- [`--secret-add` and `--secret-rm`](../reference/commandline/service_update.md#adding-and-removing-secrets) flags for `docker service update`
+
+## Example: Use secrets with a service
```
We need an example for rotation of a secret (add a new secret, remove the old one, and ensure the new secret has the same target as the old one.

We should probably also give advanced examples on how to change the permissions of the secret file, and the use of `source=bla,target=bla,uid=1000,gid=1000,mode=0700`

thaJeztah commented:
Hm, for fun I tried adding a binary as secret, and that works :smile:

```bash
$ docker run --name hola hello-world

$ docker cp hola:/hello .
$ cat hello | docker secret create hello
r4gmhiw6fmiorqfk5reqbyryr

$ docker service create --name foo --secret source=hello,target=hello,mode=0755 nginx:alpine


$ docker exec $(docker ps -n1 -q --filter label=com.docker.swarm.service.name=foo) /run/secrets/hello

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker Hub account:
 https://hub.docker.com

For more examples and ideas, visit:
 https://docs.docker.com/engine/userguide/
```

Should we document what size limits there are to secrets? (I don't think we put a size limit on there, but given that secrets are in memory, using many / big data there could have consequences perhaps?)

ehazlett commented:
Yes there is a size limit of 500 KB (https://github.com/docker/swarmkit/blob/master/manager/controlapi/secret.go#L22) per secret.  We should document this.

mstanleyjones commented:
@diogomonica @ehazlett @stevvooe @cyli @toli PTAL. I think the topic is starting to shape up nicely. I addressed the feedback so far, and added a section on rotating secrets. Still waiting on an update to the official `mysql` image so that I can use that in the example instead of `mysql/mysql-server`.

mstanleyjones commented:
cc/ @johndmulhausen @hansfrederickbrown

stevvooe commented:
LGTM

mstanleyjones commented:
Added info about binary content and 500kb limit.

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,278 @@
+decrypted secret is mounted into the container in an in-memory filesystem at
+`/run/secrets/<secret_name>`. You can update a service to grant it access to
+additional secrets or revoke its access to a given secret at any time. When the
+task container stops running, the secret is unavailable to that container. When
```
Non-blocking:  do you mean if you `docker start` the stopped container?

mstanleyjones replied:
No, I was trying to cover the case where a node is running more than one task, each of which may have access to the same secret. When one of the containers stops, it doesn't have access to the secret anymore. If none of a node's task containers need access to a given secret anymore, I think the secret is flushed from the node as well, unless the node is acting as a backup master. Right? Is there a better way to put this?

cyli replied:
I think your new change is great - I was wondering if you were specifically trying to call out doing a commit in a running container so it'd be available later, or restarting a container, or otherwise checking the contents of the stopped container on disk, etc.

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,278 @@
+
+## Example: Rotate a secret
+
+This example builds upon the previous one. In this scenario, you create a new
```
"create a new secret with..."

mstanleyjones replied:
Done

mstanleyjones commented:
Addressed @cyli feedback, tried to reword the area about when secrets get removed from a node's memory. PTAL at that section again and give me suggestions how I can improve it.

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+those containers that need access to it. A _secret_ is a blob of data that is
+encrypted during transit and at rest in a Docker swarm. A given secret is only
+accessible within those containers which have been granted explicit access to it,
+and only while those containers are running.
```
maybe we replace "accessible within those containers" to "accessible to those services"?

Other than that, beautiful.

mstanleyjones replied:
done

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+runtime but you don't want to store in the image or in source control, such as:
+
+- Usernames and passwords
+- Database names
```
database credentials

mstanleyjones replied:
I really did mean the database name, separately from the credentials (username and password). For instance, you can specify that as a file with the new `wordpress` image enhancements.

mstanleyjones replied:
Reworded a bit.

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+
+When you add a secret to the swarm, Docker sends the secret to the swarm manager
+over a mutual TLS connection. The secret is stored in the RAFT log, which is
+encrypted. The entire RAFT log is replicated across the other managers, ensuring
```
We should probably link here to another document talking about `unlock` and `unlock-key`, and how the key used for encryption is stored and how it can be managed/rotated/etc.

mstanleyjones replied:
Waiting for https://github.com/docker/docker.github.io/pull/694 to be merged.

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+secret with the new MySQL password, update the `mysql` and `wordpress` services
+to use it, then remove the old secret.
+
+1.  Create the new secret, and name it `mysql_password_2016`.
```
let's use `mysq_password_v2` instead of 2016

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+## Build support for Docker Secrets into your images
+
+If you develop a container that can be deployed as a service and accepts a
+sensitive string, such as a credential, as an environment variable, consider
```
`and requires sensitive data, such as...`

diogomonica commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+being passed directly.
+
+>**Note**: Docker secrets do not set environment variables directly. This was
+a conscious decision, since environment variables can leak between containers
```
I would do `since it's easy for environment variables to be unintentionally leaked`

mstanleyjones replied:
Done

cyli commented:
@mstanleyjones This looks great, thank you!  I just wasn't sure if you were trying to emphasize something in particular, and this removes that confusion.  Appreciate all your hard work on this!  Other than @diogomonica's suggestion for lock and unlock, this all LGTM!

ehazlett commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,280 @@
+## Read more about `docker secret` commands
+
+Use these links to read about specific commands, or continue to the
+[example about using secrets with a service](secrets.md.md@example-use-secrets-with-a-service).
```
Should this be `secrets.md@example-use-secrets-with-a-service`?

thaJeztah replied:
probably `@` -> `#` as well?

ehazlett commented:
one minor comment otherwise LGTM -- nice work!

cyli commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,340 @@
+
+    ```bash
+    $ docker secret ls
+    ID                          NAME                CREATED             UPDATED             SIZE
````
Also, many apologies for springing this on you at the last minute, but https://github.com/docker/docker/pull/28727#issuecomment-262396148 removes displaying sizes and digests at all when you inspect.  Could you delete the last column in this sample output?  (the size column?)

mstanleyjones commented:
Addressed latest feedback. Thanks @cyli !

mstanleyjones commented:
Modified the instructions in the `mysql` lab now that https://github.com/docker-library/mysql/pull/237 is merged and the `mysql` image has been updated!

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,338 @@
+    The secret is stored in the encrypted RAFT logs for the swarm.
+
+4.  Create the MySQL service. The `mysql` image now supports
+    additional environemnt variables which read their values from a file within
```
"environment" mispelling :)

cyli replied:
Also, non-blocking nitpick - this reads to me like the environment variable reads values from a file?  Whereas it's the script within the image which reads the file pointed at from the environment variable.  I'm not sure the best way to say this in a clear way, but maybe something like "additional environment variables, which lets mysql read secret values from a file..."

That may make the next sentence a bit redundant though.

cyli commented:
@mstanleyjones Thank you!  I have a minor spelling/phrasing nitpick, but everything looks great!  👍

mstanleyjones commented:
Addressed this new feedback. Thanks @cyli !

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,339 @@
+    input.
+
+    ```bash
+    $ echo "This is a secret" | docker secret create secret_data.txt
````
Wonder if we should use `secret_data.txt` as a name here, this may be confusing because either users think this is the name of a local file ("where does the `secret_data.txt` file come from?"), or that it must match the filename under which it can be accessed inside the container (which is later shown that it can be set through `target`).

Perhaps something that more obviously shows it's a custom name, such as `my-little-secret` (without file-extension)

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,339 @@
+    ```bash
+    $ docker service  create --name="redis" --secret="secret_data.txt" redis:alpine
+    ```
+
````
In this step, perhaps we should explain that (by default), the secret is accessible inside the container as a file named `/run/secrets/<name-of-your-secret>`

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,339 @@
+    ```bash
+    $ docker ps |grep redis
+
+    ea3839998cbf        redis@sha256:f1ed3708f538b537eb9c2a7dd50dc90a706f7debd7e1196c9264edeea521a86d                "docker-entrypoint..."   About a minute ago   Up About a minute   6379/tcp              redis.1.rn9sbes7lxsiaizfqzo7g4nvu
````
This output will be updated before GA to match the 1.12 output again, and the digest will no longer be shown by default, just `redis:latest`

thaJeztah replied:
(We should make this easier to do (`docker ps | grep` should not be the way, because `docker ps` does not relate to _services_))

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+
+In this example, you create a MySQL service with a custom username and password.
+Next, you add these credentials as secrets. Next, you create a WordPress
+service which uses these credentials to connect to MySQL. Finally, you update
```
Perhaps avoid the repeated "Next", something like;

```
In this example, you create a MySQL service with a custom username and password,
and add these credentials as secrets. Next, you create a WordPress
service which uses these credentials to connect to MySQL. ....
```

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,339 @@
+
+    ```bash
+    $ openssl rand -base64 20 |docker secret create mysql_password
+      lwxgks2rojdbvaui5kay1suvr
````
Can you add something to the description like "after the `secret create` command completes, the ID of the secret is printed"

Looking at this example, I can mistake `lwxgks2rojdbvaui5kay1suvr` for being the actual secret

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+    The secret is stored in the encrypted RAFT logs for the swarm.
+
+4.  Create the MySQL service. The `mysql` image now supports additional
+    environment variables which allows the container to read credentials and other
```
I'd drop "now" and "additional" here

```Markdown
The `mysql` image supports  environment variables which allows
```

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+
+4.  Create the MySQL service. The `mysql` image now supports additional
+    environment variables which allows the container to read credentials and other
+    important settings from files within the container. This example uses the new
```
Remove "new" here

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+    environment variables which allows the container to read credentials and other
+    important settings from files within the container. This example uses the new
+    `MYSQL_ROOT_PASSWORD_FILE` variable and sets it to the file path within the
+    container which contains the `mysql_password` secret.
```
Should there be a link to the readme? "For an overview of all environment variables available for configuring the mysql image, refer to the image's documentation on Docker Hub https://hub.docker.com/_/mysql/

(or Docker Store; https://store.docker.com/images/3083290a-203f-4c04-b2de-cc057959d2c9?tab=description)

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+6.  Now that MySQL is set up, you can create a WordPress container that connects
+    to the MySQL server. The WordPress service has the following characteristics:
+
+    - Joins the service to the `wp_network` overlay network, and resolves to the
```
(Same may apply to the MySQL service in point `4.`

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,339 @@
+                    --secret-rm mysql_password
+   ```
+
+4.  Verify that WordPress works by going to `http://localhost:30000/` again.
````
Hm, okay, bad news; this example shows how to replace a secret, but unfortunately it doesn't work;

- both the `wordpress` and `mysql` services use an anonymous volume; each time these services are updated, a new volume is created, so any changes made are lost. After updating, I end up with a _fresh_ database, and a _fresh_ WordPress installation (i.e., uploads, and customizations are no longer there)
- the `mysql` image does not currently have an option to _update_ an existing password through environment variables. The provided secret/password is only used during first initialization, but ignored after that


To see that it doesn't work;

Create volumes in advance for both services (so that they keep modifications);

    docker volume create wp_data
    docker volume create mysql_data


Or implicitly create the volumes during `docker service create`;

    --mount source=mysql_data,target=/var/lib/mysql
    --mount source=wordpress_data,target=/var/www/html

```bash
docker service create --name="mysql" \
                        --mount source=mysql_data,target=/var/lib/mysql \
                        --network="wp_network" \
                        --replicas=1 \
                        --secret=mysql_password \
                        -e MYSQL_ROOT_PASSWORD_FILE=/run/secrets/mysql_password \
                        mysql

docker service create --name="wordpress" \
                        --mount source=wordpress_data,target=/var/www/html \
                        --network="wp_network" \
                        --replicas="1" \
                        --publish 30000:80 \
                        --secret source=mysql_password,target=wp_db_password,mode=0400 \
                        -e WORDPRESS_DB_PASSWORD_FILE="/run/secrets/wp_db_password" \
                        -e WORDPRESS_DB_HOST="mysql:3306" \
                        wordpress:latest

```bash

Now rotate the secrets;

```bash
docker service update mysql \
                 --secret-add source=mysql_password_v2,target=mysql_password \
                 --secret-rm mysql_password

docker service update wordpress \
                --secret-add source=mysql_password_v2,target=wp_db_password,mode=0400 \
                --secret-rm mysql_password
```


After this, the mysql service is running, but still uses the old password,
the WordPress service is not running, because it cannot connect to the
database

```
MySQL Connection Error: (1045) Access denied for user 'root'@'10.0.0.3' (using password: YES)
Warning: mysqli::mysqli(): (HY000/1045): Access denied for user 'root'@'10.0.0.3' (using password: YES) in - on line 19
```

mstanleyjones replied:
Hmm, I need to test this again because it did work when I ran through my own example before...

thaJeztah replied:
It's deceiving, because if you didn't actually create a site, you'll see the default "Welcome to WordPress" screen, but it's a fresh install

diogomonica replied:
@cyli @thaJeztah ideas on how to fix this?

thaJeztah replied:
Not directly, I think updating password needs to be done through the MySQL client / CLI (some care has to be taken as well, because usernames are not unique in MySQL, multiple users with the same name can exist 😞 http://dev.mysql.com/doc/refman/5.7/en/adding-users.html)

ping @tianon any thoughts on implementing something in the MySQL image that does not require doing it "the manual way"?

cyli replied:
Agree it would have to be done manually.  I'm not sure it would make sense to have the image automatically change passwords - if database files already exist, it's sort of strange to have an image by default go in and modify it; just like it doesn't modify any existing schemas if the DB already exists - it just initializes the schema if the DB doesn't already exist.

Can we run a one-off service (with `--restart-condition=none`) that runs the same mysql container, with both secrets, and uses the CLI to use the existing passphrase to connect to the existing mysql container and set the passphrase to the new one?

tianon replied:
Yeah, I definitely agree with @cyli that having the `mysql` image itself change existing database contents automatically would be strange, although it might be interesting to make a special-purpose image for that sort of thing if someone doesn't have one already.

diogomonica replied:
I say we add a new simpler example (rotation of nginx TLS keys), and modify this mysql example to have a step where people `docker exec mysql change-password` into the running mysql on the host by using a new randomly generated password, and then add that new password as a new secret, and rotate the FE server.

This will be the way to get something like this done once exec works for swarm-mode, so we should probably just do the tutorial like this.

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,339 @@
+    - Because the scale is set to 1, only a single MySQL task runs.
+    - Grants the service access to the `mysql_password` secret, which will be
+      available as plain-text within the container at `/run/secrets/mysql_password`.
+    - Sets the environment variable `MYSQL_ROOT_PASSWORD_FILE` to this file path.
```
I think we should follow best practice here and don't use `MYSQL_ROOT_PASSWORD` for the database connection, but specify a `MYSQL_DATABASE `, `MYSQL_USER`, and `MYSQL_PASSWORD` so that a non-database-root user is used for the WordPress connection / database

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

cyli commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,351 @@
+    input.
+
+    ```bash
+    $ echo "This is a secret" | docker secret create my_secret_data
````
I also wanted to mention the change that https://github.com/docker/docker/pull/28629 is making to this command (it'd be `docker secret create my_secret_data -f -` in this particular example that passes the data via STDIN; it can also take a file name with `docker secret create my_secret_data -f <somefile>`)

diogomonica commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,351 @@
+    at `/run/secrets/mysql_password` within the container.
+
+    ```bash
+    $ docker service update mysql \
````
This will do a rolling restart of mysql, which isn't realistic for a database.

diogomonica replied:
We might want to point that out, if we want to keep it.

diogomonica replied:
I think we should add an example of certificate rotation for an nginx service. It's a lot easier/more straightforward example. Alternatively, we can show rotation of a redis access database. Running MySQL database as a service is a hard sell. We might simply show it as a normal docker run, and manual update of the passphrase using `docker exec`, and rotation of the passphrase on the wordpress using `service update`

cyli commented:
Also, https://github.com/docker/docker/pull/29061 updates `swarm leave` so that you have to provide `--force` in order to do so if the node is locked.  Not sure if that needs to be added to this doc (or if it should be updated in the `leave` docs)

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,351 @@
+`/run/secrets/<secret_name>`. You can update a service to grant it access to
+additional secrets or revoke its access to a given secret at any time.
+
+A node only has access to (encrypted) secrets if it is a swarm manager or if it
```
Should this be "decrypted" secrets?  The manager has access to all the decrypted secrets, as well as the service.

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,351 @@
+the same high availability guarantees for secrets as for the rest of the swarm
+management data.
+
+>**Warning**: RAFT data is encrypted in Docker 1.13 and higher. If any of your
```
Similarly to https://github.com/docker/docker.github.io/pull/694#discussion_r89906193, we should probably change "RAFT" to be "Raft" here and in the rest of the doc.

mstanleyjones commented:
Extensive reworking of the examples with a lot of help from @cyli. I'd also like @NathanMcCauley to have a look and make sure we are doing reasonable things here.

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,770 @@
+    $ docker service ps --no-trunc redis
+
+    NAME                               lIMAGE         NODE              DESIRED STATE  CURRENT STATE           ERROR  PORTS
+    redis.1.cezsyv6xljjj6bzpajbc6agkb  redis:alpine   ip-172-31-46-109  Running        Running 26 seconds ago
```
The output format of `docker service ps` changed back to the 1.12 format, so no `--no-trunc` is needed;

```bash
$ docker service ps web

ID            NAME   IMAGE         NODE          DESIRED STATE  CURRENT STATE          ERROR  PORTS
kht7jvzvai5z  web.1  nginx:alpine  371b354ecd17  Running        Running 3 seconds ago
```

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,770 @@
+    you would see something like this:
+
+    ```bash
+    $ docker service ps redis
````
same here

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,770 @@
+    $ docker run --rm -it committed_redis cat /run/secrets/my_secret_data
+
+    cat: can't open '/run/secrets/my_secret_data': No such file or directory
+    ```<!--'fix-->
````
^^ leftover comment `<!--'fix-->`, or intentionally?

mstanleyjones replied:
It's intentional to fix the unbalanced `'` in `can't`, otherwise the rest of the file is screwed up in Atom! I'll move it closer to the source of the problem.

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,770 @@
+    output of `docker ps`.
+
+    ```bash
+    $ docker ps --filter name=redis
````
Alternatively, use

```
docker inspect <task-id> --format '{{ .Status.ContainerStatus.ContainerID }}'
```

but that may be a bit awkward as well

mstanleyjones replied:
That seems needlessly complicated, and it seems useful to teach the reader about the --filter thing..

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,770 @@
+    ```bash
+    $ docker service rm redis
+
+    $ docker secret rm my_secret_data.txt
````
remove `.txt` here

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,770 @@
+
+### Intermediate example: Use secrets with a Nginx service
+
+**Note**: This example uses a single-Engine swarm for simplicity, and only
```
Should this be formatted as "note"?

```Markdown
> **Note**: ......
```

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,770 @@
+                            --secret nginx.conf \
+                            --publish 3000:443 \
+                            nginx:latest \
+    sh -c "ln -s /run/secrets/nginx.conf /etc/nginx/conf.d/ nginx.conf && nginx -g 'daemon off;'"
```
this should use `exec` so that nginx is PID 1 inside the container after the symlink was created;

```
sh -c "ln -s /run/secrets/nginx.conf /etc/nginx/conf.d/ nginx.conf && exec nginx -g 'daemon off;'"
```

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,770 @@
+    ID                          NAME                  CREATED             UPDATED
+    2hvoi9mnnaof7olr3z5g3g7fp   site.key       58 seconds ago      58 seconds ago
+    aya1dh363719pkiuoldpter4b   site.crt       24 seconds ago      24 seconds ago
+    zoa5df26f7vpcoz42qf2csth8   nginx.conf     11 seconds ago      11 seconds ago
```
to make it easier to follow (especially with the symlink), you could consider naming the secret (e.g.) `site.conf`

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,770 @@
+    This ensures that you don't need to rebuild your image if you change the
+    Nginx configuration.
+
+    **Note**: Normally you would create a Dockerfile which copies the `nginx.conf`
```
Should probably be formatted as `note` here as well (`> **Note**: ...`)

mstanleyjones commented:
Addressed feedback so far. Will do another pass for formatting after Netlify finishes.

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,771 @@
+
+    - Because the scale is set to `1`, only a single WordPress task runs.
+      Load-balancing WordPress is left as an exercise to the reader, because of
+      limitations with storing WordPress session data on the container
```
Should look into this again, maybe I was mistaken about this and it stores them in the database; let me know if you want me to find out

tianon replied:
I think sessions are in the database, but uploads and other theme/content related bits are definitely stored in the local volume on disk, so sharing that storage somehow is going to definitely be required for load-balancing WordPress. :+1:

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,771 @@
+      $ docker run --name mysql -d \
+                   -v $(pwd)/tempdir:/initpw \
+                   -e MYSQL_ROOT_PASSWORD_FILE=/initpw/mysql_password.txt \
+                   -p 3306:3306 \
```
I originally ran into errors attaching the container to the network, but looking it up some more it looks like I just needed to have an `--attachable` flag when creating the network?

thaJeztah replied:
If we do `docker run`, yes, but for a service not of course

cyli commented:
@thaJeztah I don't feel strongly about the container vs a service.  Regarding the gist you provided, you are right that `docker exec <mysql-container> bash -c 'mysqladmin --user=wordpress --password="$(< /run/secrets/mysql_password)" password "$(< /run/secrets/mysql_password_v2)"'` (wow your shell-fu is WAY better than mine!) doesn't seem to put anything bad anywhere.

Maybe we would also want to include a caveat about why it's ok in this case though, but that passing in secrets on the command line are generally a bad idea.  I just don't want users seeing this and thinking it's always appropriate.  For instance, in this case:

- we are not running an interactive shell, so it never goes into bash history, but even if the command were executed in an interactive shell, it would appear as `mysqladmin --user=root --password="$(< /run/secrets/mysql_password_new)" password "$(< /run/secrets/mysql_password)"` in the bash history
- `mysqladmin` masks command line params in `ps` and `/proc` - for instance, the above command is too short to capture in ps or proc, but if you run `docker exec <mysql-container> bash -c 'mysqladmin --user=wordpress --password="$(< /run/secrets/mysql_password)" password` and leave out the last password interpolation, I see the following:

    ```
    $ ps aux
    USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
    mysql        1  0.1  9.7 1328056 198804 ?      Ssl  23:24   0:00 mysqld
    root     24626  0.0  0.1  25340  3264 ?        Ss   23:39   0:00 mysqladmin --user=root --password=x x password

    $ cat /proc/24626/cmdline
    mysqladmin--user=root--password=xxpassword
    ```
 
    Not all applications do this, and so if they passed in command line args it'd be visible from viewing processes or in /proc.

Also, in the revised example, should we mention that the service update to mysql causes the container to restart and thus would cause a disruption in the wordpress service?

cyli commented:
@thaJeztah Also, 👍  on the separate `wordpress` and `mysql` users!

thaJeztah commented:
> wow your shell-fu is WAY better than mine!

"Learn by example" - I looked into some of @tianon's scripts to see how it's done 👑 

> Also, in the revised example, should we mention that the service update to mysql causes the container to restart and thus would cause a disruption in the wordpress service?

Probably good to mention.

My goals were, first of all, to avoid having password in plain text on-disk. I know a lot of work has gone into the secrets feature to avoid just _that_, and ending up with writing to plain text files in the documentation didn't feel "right".

At first, I wanted to avoid the `docker exec` as a whole, and spin up a `mysql` container to use as client (which is the "neat" way to do it). Unfortunately, we don't have secrets yet on `docker run`, so that didn't work.

As a follow up to this PR, we should do some thinking about how to actually _use_ the secrets in various situations. For example, the WordPress image still sets the secrets through environment variables, so a stray `phpinfo()` page (yup expect those to be in a lot of setups) will happily show the password;

<img width="952" alt="screen shot 2016-12-22 at 01 20 50" src="https://cloud.githubusercontent.com/assets/1804568/21410977/329b15b4-c7e5-11e6-86f2-0b92931d8145.png">

Preventing that needs changes to the configuration files of images, so we may want some examples for that.

cyli commented:
> My goals were, first of all, to avoid having password in plain text on-disk. I know a lot of work has gone into the secrets feature to avoid just that, and ending up with writing to plain text files in the documentation didn't feel "right".

Yep, don't disagree, although in theory it's on your local machine, so less terrible.  It can also be stored in a password manager such as the lastpass CLI, which would also let you pipe it into secrets, but wasn't sure if that'd be too much for an example.

> At first, I wanted to avoid the docker exec as a whole, and spin up a mysql container to use as client (which is the "neat" way to do it). Unfortunately, we don't have secrets yet on docker run, so that didn't work.

Can we use a service with `--restart-condition=none`?

```
docker service create \
  --name=mysql_password_rotation \
  --network=wpnet \
  --secret=mysql_password \
  --secret=mysql_password_v2 \
  --restart-condition=none \
  mysql \
  bash -c 'mysqladmin --host=mysql --user=wordpress --password="$(< /run/secrets/mysql_password)" password "$(< /run/secrets/mysql_password_v2)"'
```

I'm not sure if we're starting to get into weird service esoteric stuff, now, though :)

> As a follow up to this PR, we should do some thinking about how to actually use the secrets in various situations. For example, the WordPress image still sets the secrets through environment variables, so a stray phpinfo() page (yup expect those to be in a lot of setups) will happily show the password

Ah good catch.  :|  I didn't realize it'd be part of the process's environment - I'd only checked mysql's, which didn't seem to be.  cc @tianon

thaJeztah commented:
> I'm not sure if we're starting to get into weird service esoteric stuff, now, though :)

definitely taking it too far. we can revisit once (if) secrets are available for `docker run`

> Ah good catch. :| I didn't realize it'd be part of the process's environment - I'd only checked mysql's, which didn't seem to be.

Was already discussing with him; we could probably `unset` the env-vars directly after they have been consumed. Help me remind, I can try a pull request (or bake Tianon some cookies for christmas)

mstanleyjones commented:
Pushed some updates. This puts back the `mysql` service and also uses @thaJeztah magical way to update the password in MySQL. The only thing it doesn't do yet is use a non-root user for the database. Need to figure out how to do that.

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,812 @@
+                       --mount type=volume,source=mydata,destination=/var/lib/mysql \
+                       --secret source=mysql_password,target=mysql_password \
+                       -e MYSQL_ROOT_PASSWORD_FILE="/run/secrets/mysql_password" \
+                      mysql:latest
```
You need _two_ secrets here; one for the `root` password and one for the `wordpress` user;

```
docker service create \
  --name mysql \
  --replicas 1 \
  --network=mysql_private \
  --mount type=volume,source=mydata,destination=/var/lib/mysql \
  --secret=mysql_root_password \
  --secret=mysql_password \
  -e MYSQL_ROOT_PASSWORD_FILE="/run/secrets/mysql_root_password" \
  -e MYSQL_PASSWORD_FILE="/run/secrets/mysql_password" \
  -e MYSQL_DATABASE=wordpress \
  -e MYSQL_USER=wordpress \
  mysql
```

here;

- `MYSQL_ROOT_PASSWORD_FILE` sets the path to the secret for the `root` password
- `MYSQL_PASSWORD_FILE ` sets the path to the secret for the regular user's (`wordpress`) password
- `MYSQL_DATABASE` sets the name of the database to create the first time the service is started
- `MYSQL_USER` sets the name of the _user_ to create the first time the service is started

thaJeztah replied:
Make sure to use the right secret for the right user.

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,812 @@
+                     --publish 30000:80 \
+                     --secret source=mysql_password,target=wp_db_password,mode=0400 \
+                     -e WORDPRESS_DB_PASSWORD_FILE="/run/secrets/wp_db_password" \
+                     -e WORDPRESS_DB_HOST="mysql:3306" \
```
You need to add `WORDPRESS_DB_USER` here, to set the name to use for connecting to the MySQL database;

```
-e WORDPRESS_DB_USER=wordpress \
```

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,812 @@
+                     --replicas 1 \
+                     --network mysql_private \
+                     --publish 30000:80 \
+                     --secret source=mysql_password,target=wp_db_password,mode=0400 \
```
Add a volume here, so that the WordPress data is preserved when updating the service;

```
--mount type=volume,source=wpdata,destination=/var/www/html \
```

mstanleyjones replied:
What WordPress data is that? I thought it stored everything in the MySQL database....

thaJeztah replied:
The WordPress installation (theme, uploads, configuration, etc)

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,812 @@
+    ```bash
+    $ docker service rm wordpress
+
+    $ docker rm -f mysql
````
this will be `docker service rm mysql`, or combine with the previous one (`docker service rm wordpress mysql`)

also, the named volumes can be cleaned up afterwards

mstanleyjones commented:
OK, the latest update uses a separate `wordpress` database, user, and secret. I still need to set the the MySQL root password, but it's not used after MySQL starts up.

This is getting very close to being ready to go, IMHO. I guess the only outstanding concern is the WordPress environment variable, right @thaJeztah ?

mstanleyjones commented:
BTW I don't need to use `WORDPRESS_DB_USER` because it defaults to `wordpress`.

thaJeztah commented:
> I guess the only outstanding concern is the WordPress environment variable

That's something to be addressed in the official WordPress image; I haven't had time to look into that yet, but it's orthogonal to this PR

> BTW I don't need to use WORDPRESS_DB_USER because it defaults to wordpress.

Oh, yes, it probably defaults to that; doesn't hurt to be explicit, but I'm also fine to skip setting it

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,857 @@
+    l1vinzevzhj4goakjap5ya409
+    ```
+
+    The value returned is not the password, but the ID (digest) of the secret.
````
small nit; remove `(digest)` here; the ID of secrets have no relation to the content of the secrets (otherwise they may potentially "leak" the content).

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,857 @@
+
+    - Because the scale is set to `1`, only a single MySQL task runs.
+      Load-balancing MySQL is left as an exercise to the reader and involves
+      than just scaling the service.
```
nit: missing "more" before "than"

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,857 @@
+      initializing the system database for the first time. Afterward, the
+      passwords are stored in the MySQL system database itself.
+    - Sets environment variables `MYSQL_USER` and `MYSQL_DATABASE`. A new
+      database called `wordpress` is created when the image starts, and the
```
nit: "when the image starts" -> "when the container starts"

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,857 @@
+    wvnh0siktqr3  mysql  replicated  1/1       mysql:latest
+    ```
+
+    At this point, you could actually revoke the `mysql` service's access to the
````
👍

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,857 @@
+Rotating passwords or other secrets will often involve additional steps outside
+of Docker.
+
+1.  Create the new password and store it in file in the `tempdir` directory
```
remove "and store it in file in the ..." part, as it's no longer needed

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,857 @@
+
+    ```bash
+    $ docker service update \
+                     --secret-rm mysql_password mysql
````
This is something we need to fix; doing `--secret-rm` and `--secret-add` should be possible in a single update

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,857 @@
+    First, find the ID of the `mysql` container task.
+
+    ```bash
+    $ docker ps | grep mysql |awk {'print $1'}
````
It's better to use the built-in filter functionality here (and `-q` to only show the `ID`;

```
docker ps -q --filter name=mysql
```

thaJeztah commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,857 @@
+    Verify that the blog post you wrote still exists, and if you changed any
+    configuration values, verify that they are still changed.
+
+6.  Revoke accsss to the old secret from the MySQL service and
```
typo `accsss`

mstanleyjones commented:
Fixed latest problems found by @thaJeztah and also got a review from my friend @bskaggs as well. He spotted things I could no longer see, I've been looking at this too long!

thaJeztah commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,857 @@
+    ```bash
+    $ docker service rm wordpress mysql
+
+    $ docker volume rm mydata
````
need to add `wpdata` here as well now

mstanleyjones commented:
@thaJeztah nit fixed, and a few more picky formatting tweaks. I think this is ready to merge but I'd like input from @cyli @diogomonica @NathanMcCauley

cyli commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,837 @@
+    `service update` command redeploys the service.
+
+    ```bash
+    $ docker exec -it <container_id> cat /run/secrets/my_secret_data
````
Non-blocking:  since we had the example above of `docker exec $(docker ps --filter name=redis -q) cat /run/secrets/my_secret_data`, can we use that same command here instead of `<container_id>`?

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+
+### Intermediate example: Use secrets with a Nginx service
+
+> **Note**: This example uses a single-Engine swarm for simplicity, and only
```
Non-blocking:  Since all of these examples (redis, nginx, and wordpress) presume a single-Engine swarm, should that part of this note be moved to the top of the examples section?  (Similarly for the wordpress example)

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+
+    > **Note**: Normally you would create a Dockerfile which copies the `site.conf`
+    > into place, build the image, and run a container using your custom image.
+    > This example does it all in one step.
```
Maybe something like "This example does not require a custom image; it copies the `site.conf` into place and runs the container all in one step"?

cyli commented on engine/swarm/secrets.md:
````diff
@@ -0,0 +1,837 @@
+    yvsczlx9votfw3l0nz5rlidig   mysql_root_password   12 seconds ago      12 seconds ago
+    ```
+
+    The secrets are stored in the encrypted WAL logs for the swarm.
````
Non-blocking:  Should we just say "Raft logs" here, since above we mention that secrets are stored in the raft logs?  It's true that they are written to the WAL (and snapshot, later on), but that might be too much raft detail for this document, particularly since we never say what WAL stands for.

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+> a file on disk. You must use a query or a `mysqladmin` command to change the
+> password in MySQL.
+
+1.  Generate a random alphanumeric password for MySQL and store it as a Docker
```
Should we mention that this first password is for the wordpress user for mysql?  Or at least a non-root user?

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+queries or commands, as opposed to just changing a single environment variable
+or a file, since the image only sets the MySQL password if the database doesn’t
+already exist, and MySQL stores the password within a MySQL database by default.
+Rotating passwords or other secrets will often involve additional steps outside
```
Nonblocking nitpick: should this be "may"?  If it's something simple like an API key the service just access, just changing the service spec should be fine.  Alternately, this is likely always the case with databases, so maybe we can call out DBs specifically here?

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+    `/run/secrets/mysql_password`.
+
+    Even though the MySQL service has access to both the old and new secrets
+    now, the MySQL root password has not yet been changed.
```
We don't change the root password at all in this example, right?  Could we clarify this example to say that we don't need to adjust the service's access to the MySQL root password, since it's not being modified?

cyli commented on engine/swarm/secrets.md:
```diff
@@ -0,0 +1,837 @@
+    Even though the MySQL service has access to both the old and new secrets
+    now, the MySQL root password has not yet been changed.
+
+3.  Now, change the MySQL password for the `wordpress` user using the `mysql`
```
"`mysqladmin` CLI", as opposed to "`mysql` CLI"

mstanleyjones commented:
Addressed @cyli feedback, and also pushed a second commit which updates the syntax of the `-f` flag to `docker create`, as per https://github.com/docker/docker/pull/29889.

cyli commented:
LGTM!
